### PR TITLE
Fix `declaration-block-no-redundant-longhand-properties` autofix conflicts

### LINKED ISSUE #7610: Fix `declaration-block-no-redundant-longhand-properties` autofix conflicts
### What minimal example or steps are needed to reproduce the bug?

```css
a {
	border-top-width: foo;
	border-right-width: foo;
	border-bottom-width: foo;
	border-left-width: foo;

	border-top-color: foo;
	border-right-color: foo;
	border-bottom-color: foo;
	border-left-color: foo;

	border-top-style: foo;
	border-right-style: foo;
	border-bottom-style: foo;
	border-left-style: foo;
}
```

### What minimal configuration is needed to reproduce the bug?

```json
{
  "rules": {
    "declaration-block-no-redundant-longhand-properties": true
  },
  "fix": true
}
```

### How did you run Stylelint?

https://stylelint.io/demo

### What did you expect to happen?

```css
a {
	border-width: foo foo foo foo;
	border-color: foo foo foo foo;
	border-style: foo foo foo foo;
}
```
or
```css
a {
	border-top: foo foo foo;
	border-right: foo foo foo;
	border-bottom: foo foo foo;
	border-left: foo foo foo;
}
```

### What actually happened?

demo

```css
a {
	border-width: foo foo foo foo;
	border-color: foo foo foo foo;
}
```

locally

```css
a {
	border-width: foo foo foo foo;
	border-color: foo foo foo foo;
    border-left-style: foo;
}
```

@ybiquitous @jeddy3
It seems to me that it's a **major bug**.

### PR DESCRIPTION
**Pull request by Mouvedia**
> Which issue, if any, is this issue related to?

Closes #7610

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

### CODE REVIEW

File: lib/rules/declaration-block-no-redundant-longhand-properties/index.mjs [hunk 1]
```diff
@@ -314,9 +314,14 @@ const rule = (primary, secondaryOptions, context) => {
 								});
 
 								firstDeclNode.replaceWith(newShorthandDeclarationNode);
-
 								declNodes.forEach((node) => node.remove());
 
+								longhandDeclarations.forEach((v, k) => {
+									const array = v.filter((element) => !longhandDeclaration.includes(element));
+
+									longhandDeclarations.set(k, array);
```
Comment (Mouvedia) [inline lib/rules/declaration-block-no-redundant-longhand-properties/index.mjs:322]: I could have done `v.splice(0, v.length,...v.filter((element) => !longhandDeclaration.includes(element)));` instead.

File: lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs [hunk 2]
```diff
@@ -142,6 +142,78 @@ testRule({
 			endLine: 1,
 			endColumn: 138,
 		},
+		{
+			code: `a {
```
Comment (ybiquitous) [inline lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs:146]: [suggest] I recommend using `stripIndent` like other test codes:

```js
import { stripIndent } from 'common-tags';
```

```suggestion
			code: stripIndent`a {
```

If doing so, we won't need to care about `column` values.

File: lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs [hunk 3]
```diff
@@ -142,6 +142,78 @@ testRule({
...
+				border-width: 1px 2px 3px 4px;
+				border-color: red green blue yellow;
+				border-style: dotted solid dashed double;
+			}`,
+			warnings: [
+				{
+					column: 5,
+					endColumn: 22,
+					endLine: 5,
+					line: 5,
```
Comment (ybiquitous) [inline lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs:170]: [suggest] The order of `line`, `column`, `endLine`, and `endColumn` is more consistent:

```suggestion
					line: 5,
					column: 5,
					endLine: 5,
					endColumn: 22,
```

File: lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs (hunk 2 above)
Comment (Mouvedia) [reply lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs:146]: It didn't change `endColumn` and `column`.

Comment (Mouvedia) [reply lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs:146]: It didn't change `endColumn` and `column`.
What's the desired outcome? I don't get it.

Comment (ybiquitous) [reply lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs:146]: Sorry, I mean the following should be:

```js
			code: stripIndent`
				a {
					border-top-width: 1px;
					border-right-width: 2px;
					border-bottom-width: 3px;
					border-left-width: 4px;
					border-top-color: red;
					border-right-color: green;
					border-bottom-color: blue;
					border-left-color: yellow;
					border-top-style: dotted;
					border-right-style: solid;
					border-bottom-style: dashed;
					border-left-style: double;
				}
			`,
			fixed: stripIndent`
				a {
					border-width: 1px 2px 3px 4px;
					border-color: red green blue yellow;
					border-style: dotted solid dashed double;
				}
			`,
```

File: lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs (hunk 3 above)
Comment (ybiquitous) [reply lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs]: 2c3d386 doesn't resolve it. Please see the order in the same file: https://github.com/stylelint/stylelint/blob/2c3d38652c3b9a8be6abf829ab02c45ac5bcd8f7/lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs#L95-L98

File: .changeset/large-olives-press.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+"stylelint": patch
+---
+
+Fix `declaration-block-no-redundant-longhand-properties` autofix conflicts
```
Comment (ybiquitous) [inline .changeset/large-olives-press.md:5]: [suggest] Let's clarify the resolve bug:

```suggestion
Fix `declaration-block-no-redundant-longhand-properties` autofix conflicts for `border-*`
```

**Comment by ybiquitous** [comment]
I notice that this problem is about `border-*-style` false positive, rather than generic autofix. See the [demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdugLgIwgCdZCBaHCAB1IHcBLKHAC3hQEZKAPAbi1wOIwyhOgHMmOWg2asATF17Z8REqQI4KAWymMWKAMwK+ywWQA2MAGaT6u1gBYjSgaorVIZoq0Iwoi-ipCpCLikh5eKKI+MJj+JqrqWqThhKx4ZgCuMHEuQRbWyRCeqSgAnjBmnjQ5gWRupADOOKUWrFAQGr41psFiEo3NrSgNRQzdCR1JTS0wbUgNTF3GueZWktND7Rnp2VgAviAANCCWdBYAckiacIgwnFeUFgB0YA0NR+AQmKeiCCAYmBQKHQIDuOBiUAaINYAG0sECgSCNhU6Jgwl8fgMkJgoEhiCD4SgALr7D6Qb5iABiRE0SBwfwAViNMB9YJR3ogAYiQMizKicNDgSAzHSYE0QYdCUjBii0YUKaIsTi8VBBSCReDxSBSXsgA):

image

`border-*-style` props are reported too much, comparing `border-*-width` and `border-*-color`.

File: .changeset/large-olives-press.md (hunk 4 above)
Comment (ybiquitous) [reply .changeset/large-olives-press.md:5]: See https://github.com/stylelint/stylelint/pull/7626#issuecomment-2065876959. The following may be more accurate:

```suggestion
Fix `declaration-block-no-redundant-longhand-properties` false positives for `border-*-style`
```

Comment (Mouvedia) [reply .changeset/large-olives-press.md:5]: This is the only example that I could find.
The underlying question is, if we care about perf, should we put an `if` around the `forEach`?
i.e. `if (blacklist.includes(foo)) {…}`

**Comment by Mouvedia** [comment]
Apart from

> `Expected shorthand property "border"
[(declaration-block-no-redundant-longhand-properties)](https://stylelint.io/user-guide/rules/declaration-block-no-redundant-longhand-properties)
[13:2-19]`

which will be covered by #7609 in a particular case:

```css
a {
	border-top-width: foo;
	border-right-width: foo;
	border-bottom-width: foo;
	border-left-width: foo;
	border-top-color: bar;
	border-right-color: bar;
	border-bottom-color: bar;
	border-left-color: bar;
	border-top-style: qux;
	border-right-style: qux;
	border-bottom-style: qux;
	border-left-style: qux;
}
```

the rest looks OK.
i.e. I expect 7 warnings but only 3 fixes attempted.

File: .changeset/large-olives-press.md (hunk 4 above)
Comment (Mouvedia) [reply .changeset/large-olives-press.md:5]: > for `border-*`

It's inaccurate. Only a subset of `border-*` are affected and the current implementation doesn't differentiate.
i.e. it's future proof but probably could use a blacklist instead

**Comment by Mouvedia** [comment]
@ybiquitous I have added an `if` around the `forEach` so that it doesn't loop for other shorthands.
I feel that this is a major bug and that this PR is fine as is.
If you have suggestions for other tests Ill welcome them.
Concerning a refactor, this will have to wait.

File: lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs (hunk 3 above)
Comment (Mouvedia) [reply lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs]: Sorry about that I must read it incorrectly.

File: .changeset/large-olives-press.md (hunk 4 above)
Comment (Mouvedia) [reply .changeset/large-olives-press.md:5]: I have added the blacklist in the last commit.

**Comment by ybiquitous** [comment]
@Mouvedia I'm against adding this workaround, and again we should reconsider the implementation even if it would be a big change. Honestly, I don't want to review PRs so many times only for one rule, and this workaround may occur other problems.

In addition, this bug has been not found for a long time. You say this is "major", but I guess many people are not in trouble.

I think it's a good timing to detect the rule's implementation design.

**Comment by Mouvedia** [comment]
Ill just explain what this fix does then, maybe that will change your mind.
The rule creates an array that it populates with longhands that it encounters inside a declaration block, for each shorthand supported. Once the array matches it can move on to the fixer.
Before the fix, once the replacement of the longhands by the shorthand was done it didn't remove those longhands from the other arrays that were being populated at the same time. Inevitably it would cause useless autofixes to be attempted because the arrays weren't cleaned up. The first iteration of the fix was doing the cleanup without any regard of whether the shorthands were affected by the bug. The last iteration only does it for a shortlist of shorthands, it's less future proof and requires to maintain a blacklist.

I hope that's clearer and I managed to convince you.

File: .changeset/large-olives-press.md (hunk 4 above)
Comment (ybiquitous) [inline .changeset/large-olives-press.md:5]:
```suggestion
Fixed: `declaration-block-no-redundant-longhand-properties` autofix conflicts
```

**Review by ybiquitous — APPROVED** [review]
Okay, I agree with merging this since it resolves at lease autofix problem. But I still have the opinion to reconsider this rule's implementation as soon as possible.

If you fix the changelog, you can merge. LGTM.

**Comment by ybiquitous** [comment]
The discussions about this rule have been opened in #7630 and #7631. They indicates there is some problem in the rule implementation.

**Comment by Mouvedia** [comment]
> …you can merge…

I have added more tests to cover more permutations.

> The discussions about this rule have been opened in https://github.com/stylelint/stylelint/issues/7630 and https://github.com/stylelint/stylelint/issues/7631. They indicates there is some problem in the rule implementation.

We can continue the discussion around maintenance and refactors there.

**Pull request merged**